assets/js/util/report-validation.js

```diff
@@ -0,0 +1,134 @@
+ * an array of string, an array of objects or mix of strings and objects. Objects
+ * must have "expression" and "alias" properties in order to be considered as valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} metrics The metrics to check.
+ * @return {boolean} TRUE if metrics are valid, otherwise FALSE.
+ */
+export function isValidMetrics( metrics ) {
+	return isValidStringsOrObjects( metrics, ( metric ) => {
+		const validExpression = metric.hasOwnProperty( 'expression' ) && typeof metric.expression === 'string';
+		const validAlias = metric.hasOwnProperty( 'alias' ) && typeof metric.alias === 'string';
+		return validExpression && validAlias;
+	} );
+}
+
+/**
+ * Verifies provided dimensions to make sure it matches allowed values. It can be a string,
+ * array of strings, an object with "name" field, array of such objects or an array of strings
+ * and objects.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} dimensions The dimensions to check.
+ * @return {boolean} TRUE if dimensions are valid, otherwise FALSE.
+ */
+export function isValidDimensions( dimensions ) {
+	return isValidStringsOrObjects( dimensions, ( dimension ) => {
+		return dimension.hasOwnProperty( 'name' ) && typeof dimension.name === 'string';
+	} );
+}
+
+/**
+ * Verifies that either date range or start and end dates are valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} dateRange The date range to check.
+ * @param {string} startDate The start date to check.
+ * @param {string} endDate The end date to check.
+ * @return {boolean} TRUE if either date range or start/end dates are valid, otherwise FALSE.
+ */
+export function isValidDateRange( dateRange, startDate, endDate ) {
+	const validStartDate = startDate && startDate.match( /^\d{4}-\d{2}-\d{2}$/ );
+	const validEndDate = endDate && endDate.match( /^\d{4}-\d{2}-\d{2}$/ );
+	const validDateRange = dateRange && dateRange.match( /^last-\d+-days$/i );
+
+	return ( validStartDate && validEndDate ) || validDateRange;
+}
+
+/**
+ * Verifies that order definitions are valid. It can be either an object or an array
+ * of objects where each object has "fieldName" and valid "sortOrder" properties.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object|Object[]} orders The order definitions to check.
+ * @return {boolean} TRUE if order definitions are valid, otherwise FALSE.
+ */
+export function isValidOrders( orders ) {
+	const isValidOrder = ( order ) => {
+		const isValidFieldName = order.hasOwnProperty( 'fieldName' ) && !! order.fieldName;
+		const isValidSortOrder = order.hasOwnProperty( 'sortOrder' ) && order.sortOrder.toString().match( /(ASCENDING|DESCENDING)/i );
+		return isValidFieldName && isValidSortOrder;
+	};
+
+	if ( Array.isArray( orders ) ) {
+		for ( let i = 0; i < orders.length; i++ ) {
+			if ( typeof orders[ i ] !== 'object' || ! isValidOrder( orders[ i ] ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
```

Comment:
Again: this could be refactored to use `every` rather than the `for` loop, which is just a bit more concise and applicable given we aren't operating on the array but just verifying every element passes a test.

Reply:
Updated.